### DISCUSSION TITLE
Why the "changed" event does not submit event detail in its handler?

### DISCUSSION THREAD

**Discussion by alperen** [fabricjs/fabric.js#6890] [Q&A]
I'm trying to add keyboard shortcuts into the `fabric.textbox` object to format it, I would like to listen to textbox's `changed` event to catch keyboard actions but the event does not publish any detail about the actual browser event. fabricjs should pass the keyboard event to `changed` events' handler in my opinion.

For example, we could change the firing event like this:

```js
fire('changed', hiddenTextAreaKeyupEvent)
```

`hiddenTextAreaKeyupEvent`  variable will refer to  `KeyboardEvent`.

Or is there any solution to listen to keyboard event? :)

Thanks.

**Comment by asturur** [comment]
You are right we are not sending it. We could add it, is not bad.
That would be `fire('changed', { e: e, target: this })`.
I m not sure if this will catch any kind of inputs with keyboard shortcuts, i'm also not sure how fabricJS behave if you press , for example, CMD + B while editing.

You could open a PR to add this information, and if possible a test.

**Reply by alperen** [reply]
Hmm, I'm going to try to implement it, also I'll write the test cases. After I posted that discussion, for the quick fix, I started to listen to `hiddenTextarea` then it worked for me, I could catch keyboard shortcuts via listing to `hiddenTextarea` but it was the hacky solution.

See you then.

**Reply by alperen** [reply]
How about firing `changed` event on `keydown` or something like it? Because the `InputEvent` has no detail about meta keys? And it seems it is very difficult via testing with QUnit :)

**Reply by asturur** [reply]
FabricJS does not support keyboard events, the only way to do that is indeed adding a listener to hiddenTextarea that is a good solution too.
We try to do not get in the real of application specific functionalities, as we do not provide buttons we do not provide keyboard shortcuts.

**Reply by alperen** [reply]
Yes, I see. Providing buttons or shortcuts is out of the `fabric`'s responsibility as you mention. I meant, could we fire the `changed`event on `itext`'s `onKeydown` event, thus, the event could be `KeyboardEvent` and we could get all details about `hiddenTextarea`'s events.

**Reply by asturur** [reply]
but we do not handle text changes on keydown, the text changes gets handled on the onInput function, and that is where the event fires. 
You can attach the events on hiddenTextarea using the enter:edit event, ( since everytime the hiddentextare is new ) and use keydown/keyup as you prefer.

**Reply by alperen** [reply]
Yes you're right. Actually, change and keydown are different events.
